File: packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthenticationHandler.java
package io.capawesome.capacitorjs.plugins.firebase.authentication;

import static io.capawesome.capacitorjs.plugins.firebase.authentication.FirebaseAuthenticationHelper.createSignInResult;

import android.util.Log;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.google.firebase.auth.AdditionalUserInfo;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;

public class FirebaseAuthenticationHandler {

    public enum AuthType {
        SIGN_IN,
        LINK
    }

    public static void success(
        final PluginCall call,
        final AuthType authType,
        final FirebaseAuthentication implementation,
        @Nullable AuthCredential credential,
        @Nullable String idToken,
        @Nullable String nonce,
        @Nullable String accessToken,
        @Nullable AdditionalUserInfo additionalUserInfo
    ) {
        boolean skipNativeAuth = implementation.getConfig().getSkipNativeAuth();
        if (skipNativeAuth) {
            JSObject linkResult = createSignInResult(null, credential, idToken, nonce, accessToken, additionalUserInfo);
            call.resolve(linkResult);
            return;
        }
        (
            (authType == AuthType.LINK)
                ? implementation.getCurrentUser().linkWithCredential(credential)
                : implementation.getFirebaseAuthInstance().signInWithCredential(credential)
        ).addOnCompleteListener(
                implementation.getPlugin().getActivity(),
                task -> {
                    if (task.isSuccessful()) {
                        Log.d(FirebaseAuthenticationPlugin.TAG, "linkWithCredential succeeded.");
                        AuthResult authResult = task.getResult();
                        JSObject linkResult = createSignInResult(
                            authResult.getUser(),
                            authResult.getCredential(),
                            idToken,
                            nonce,
                            accessToken,
                            authResult.getAdditionalUserInfo()
                        );
                        call.resolve(linkResult);
                    } else {
                        Log.e(FirebaseAuthenticationPlugin.TAG, "linkWithCredential failed.", task.getException());
                        call.reject(task.getException().getLocalizedMessage());
                    }
                }
            );
    }

    public static void failure(final PluginCall call, String message, Exception exception) {
        if (message == null && exception != null) {
            message = exception.getLocalizedMessage();
        }
        call.reject(message, exception);
    }
}

File: packages/authentication/ios/Plugin/FirebaseAuthenticationHandler.swift
import Foundation
import Capacitor
import FirebaseAuth

public enum AuthType {
    case signIn
    case link
}

@objc public class FirebaseAuthenticationHandler: NSObject {
    static func success(_ call: CAPPluginCall, _ authType: AuthType, _ implementation: FirebaseAuthentication, credential: AuthCredential, idToken: String?, nonce: String?, accessToken: String?) {
        self.success(call, authType, implementation, credential: credential, idToken: idToken, nonce: nonce, accessToken: accessToken, displayName: nil)
    }

    static func success(_ call: CAPPluginCall, _ authType: AuthType, _ implementation: FirebaseAuthentication, credential: AuthCredential, idToken: String?, nonce: String?, accessToken: String?, displayName: String?) {
        if implementation.getConfig().skipNativeAuth == true {
            let result = FirebaseAuthenticationHelper.createSignInResult(credential: credential, user: nil, idToken: idToken, nonce: nonce,
                                                                         accessToken: accessToken, additionalUserInfo: nil, displayName: displayName)
            call.resolve(result)
            return
        }
        (
            (authType == AuthType.link)
                ? implementation.getCurrentUser()!.link
                : Auth.auth().signIn
        )(credential) { (authResult, error) in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            let result = FirebaseAuthenticationHelper.createSignInResult(credential: authResult?.credential, user: authResult?.user,
                                                                         idToken: idToken, nonce: nonce, accessToken: accessToken,
                                                                         additionalUserInfo: authResult?.additionalUserInfo, displayName: displayName)
            call.resolve(result)
        }
    }

    static func failure(_ call: CAPPluginCall, message: String?, error: Error?) {
        let errorMessage = message ?? error?.localizedDescription ?? ""
        call.reject(errorMessage, nil, error)
    }
}
